src/common/form/index.js

```diff
@@ -158,39 +161,67 @@ const DEFAULT_UNIT = 'GiB'
   placeholder: propTypes.string,
   readOnly: propTypes.bool,
   required: propTypes.bool,
-  style: propTypes.object
+  style: propTypes.object,
+  value: propTypes.number
 })
-export class SizeInput extends Component {
+export class SizeInput extends BaseComponent {
   constructor (props) {
     super(props)
 
-    const humanSize = props.defaultValue && formatSizeRaw(props.defaultValue)
-    this._defaultValue = humanSize && humanSize.value
-    this.state = { unit: humanSize ? humanSize.prefix + 'B' : props.defaultUnit || DEFAULT_UNIT }
+    this.state = this._createState(formatSizeRaw(firstDefined(props.value, props.defaultValue, 0)))
   }
 
   componentWillReceiveProps (newProps) {
-    this.value = newProps.defaultValue
+    const { value } = newProps
+    if (value !== this.props.value && value !== this.state.bytes || 0) {
+      this.setState({
+        ...this._createState(formatSizeRaw(value)),
+        bytes: undefined
+      })
+    }
+  }
+
+  _createState = humanSize => {
+    const { defaultUnit } = this.props
+    return {
+      unit: humanSize && humanSize.value ? humanSize.prefix + 'B' : defaultUnit || DEFAULT_UNIT,
+      value: humanSize ? round(humanSize.value, 3) : ''
+    }
   }
 
   get value () {
-    const value = this.refs.value.value
-    return value ? parseSize(value + ' ' + this.state.unit) : undefined
+    const { unit, value } = this.state
+    return value ? parseSize(value + ' ' + unit) : undefined
   }
 
   set value (newValue) {
-    const humanSize = newValue && formatSizeRaw(newValue)
-    this.refs.value.value = humanSize ? humanSize.value : ''
-    this.setState({ unit: humanSize ? humanSize.prefix + 'B' : DEFAULT_UNIT })
+    if (
+      process.env.NODE_ENV !== 'production' &&
+      this.props.value != null
+    ) {
+      throw new Error('cannot set value of controlled SizeInput')
+    }
+    this.setState(this._createState(
+      formatSizeRaw(newValue || 0)
+    ))
   }
 
   _onChange = () =>
     this.props.onChange && this.props.onChange(this.value)
 
-  _updateUnit = unit => {
-    this.setState({ unit })
-    this._onChange()
+  _updateValue = event => {
+    const { value } = event.target
+    if (this.props.value != null) {
+      this.setState({
+        value,
+        bytes: parseSize((value || 0) + ' ' + this.state.unit)
+      }, () => this._onChange())
```

Comment:
Do not update the `input` from inside, what you need to do is:

1. store the requested value and the corresponding number of bytes
2. call `onChange` with the next value
3. when the `value` prop changes, check whether it equals the requested number of bytes and …